provides.py

```diff
@@ -2,29 +2,18 @@
 
 from charmhelpers.core import hookenv
 from charms.reactive import Endpoint
-from charms.reactive import when_any, when_not
-from charms.reactive import set_state, remove_state
+from charms.reactive import toggle_flag, is_flag, clear_flag, set_flag
 
 
 class CNIPluginProvider(Endpoint):
-
-    @when_any('endpoint.{endpoint_name}.changed')
-    def changed(self):
-        ''' Set the connected state from the provides side of the relation. '''
-        set_state(self.expand_name('{endpoint_name}.connected'))
-        if self.config_available():
-            set_state(self.expand_name('{endpoint_name}.available'))
-        else:
-            remove_state(self.expand_name('{endpoint_name}.available'))
-        remove_state(self.expand_name('{endpoint_name}.configured'))
-        remove_state(self.expand_name('endpoint.{endpoint_name}.changed'))
-
-    @when_not('endpoint.{endpoint_name}.joined')
-    def broken_or_departed(self):
-        '''Remove connected state from the provides side of the relation. '''
-        remove_state(self.expand_name('{endpoint_name}.connected'))
-        remove_state(self.expand_name('{endpoint_name}.available'))
-        remove_state(self.expand_name('{endpoint_name}.configured'))
+    def manage_flags(self):
+        toggle_flag(self.expand_name('{endpoint_name}.connected'),
+                    self.is_joined)
+        toggle_flag(self.expand_name('{endpoint_name}.available'),
+                    self.config_available())
```

Comment:
I was going to suggest including `self.is_joined` because it does more what you expect during the final departing hook, but since `config_available()` uses `goal_state()` I guess it's kind of moot.